.NET Aspire release 9.3 content (#3926)
dotnet/docs-aspire: docs/fundamentals/build-container-images.md

These APIs give you fine-grained control over the image building process and provide real-time feedback to users during lengthy build operations.

> [!IMPORTANT]
> These APIs are currently in preview and subject to change. They are designed for advanced scenarios where you need custom control over container image building and progress reporting. To suppress warnings for these APIs, see [Compiler Error ASPIREPUBLISHERS001](../diagnostics/aspirepublishers001.md).

## When to use these APIs

Consider using the container image building and progress reporting APIs in these scenarios:

- **Custom deployment targets**: When you need to deploy to platforms that require specific image formats or build configurations.
- **Complex build pipelines**: When your publishing process involves multiple steps that users need visibility into.
- **Enterprise scenarios**: When you need custom progress reporting for integration with CI/CD systems or dashboards.
- **Custom resource types**: When implementing custom resources that need to participate in the publishing and deployment process.

> [!NOTE]

> > For most standard Aspire applications, the built-in publishing process handles container image building automatically without requiring these APIs.

Reviewer: ```suggestion
> For most standard Aspire applications, the built-in publishing process builds container images automatically without requiring these APIs.
```

- Reuse an existing image.
- Build from a .NET project using `dotnet publish /t:PublishContainer`.
- Build from a Dockerfile using the local container runtime.

### Container build options

<!-- TODO: ContainerBuildOptions xref -->

The `ContainerBuildOptions` class, which provides strongly typed configuration for container builds. This class allows you to specify:

- **Image format**: Docker or Open Container Initiative (OCI) format.
- **Target platform**: Linux x64, Windows, ARM64, etc.
- **Output path**: Where to save the built images.

### Container runtime health checks

The builder performs container runtime health checks (Docker/Podman) only when at least one resource requires a Dockerfile build. This change eliminates false-positive errors in projects that publish directly from .NET assemblies. If the container runtime is required but unhealthy, the builder throws an explicit `InvalidOperationException` to surface the problem early.

## Publishing activity reporter API

> The `PublishingActivityProgressReporter` API enables structured progress reporting during [aspire publish](../cli-reference/aspire-publish.md) and [aspire deploy](../cli-reference/aspire-deploy.md) commands. This reduces uncertainty during long-running operations and surfaces failures early.

Reviewer: ```suggestion
The `PublishingActivityProgressReporter` API enables structured progress reporting during [`aspire publish`](../cli-reference/aspire-publish.md) and [`aspire deploy`](../cli-reference/aspire-deploy.md) commands. This reduces uncertainty during long-running operations and surfaces failures early.
```

- Uses the `PublishingActivityProgressReporter` API to create and manage deployment steps and tasks.
- Reports progress and marks each deployment phase as completed.
- Completes the deployment operation with a final status update.
- Handles cancellation through the provided `CancellationToken`.

## Best practices

When using these APIs, follow these guidelines:

### Image building

- Always specify explicit `ContainerBuildOptions` for production scenarios.
- Consider target platform requirements when building for deployment.
- Use OCI format for maximum compatibility with container registries.
- Handle `InvalidOperationException` when container runtime health checks fail.

### Progress reporting

- Encapsulate long-running logical phases in steps rather than emitting raw tasks.
- Keep titles concise (under 60 characters) as the CLI truncates longer strings.
- Call `CompletePublishAsync` exactly once per publishing or deployment operation.
- Treat warnings as recoverable and allow subsequent steps to proceed.
- Treat errors as fatal and fail fast with clear diagnostics.
- Use asynchronous, cancellation-aware operations to avoid blocking event processing.

### State management

> - Each step and task starts in *Running* state and transitions exactly once to *Completed*, *Warning*, or *Error*.

Reviewer: ```suggestion
- Ensure each step and task starts in *Running* state and transitions exactly once to *Completed*, *Warning*, or *Error*.
```

> - Attempting multiple state transitions throws an exception.

Reviewer: ```suggestion
- Throw an exception when attempting multiple state transitions.
```

- The reporter guarantees ordered events and prevents interleaving.

> - Disposal of `IPublishingActivityStep` automatically completes unfinished steps.

Reviewer: ```suggestion
- Dispose of `IPublishingActivityStep` to automatically complete unfinished steps.
```
